src/relax/transform/alter_op_impl.cc

```diff
@@ -0,0 +1,312 @@
+    }
+    return arr_tensor_sinfo;
+  }
+
+  Expr TransformLayout(const Expr& expr, const IndexMap& index_map) {
+    ObjectPtr<LayoutTransformAttrs> attrs = make_object<LayoutTransformAttrs>();
+    // We want to avoid two layout_transform ops to share the same index map even if they are
+    // identical. The scope of vars used in index map initial indices is local to the op. Not doing
+    // so would confuse the structural equality check.
+    attrs->index_map = std::move(DeepCopyIndexMap(index_map));
+    return Call(layout_transform_op_, {expr}, Attrs{std::move(attrs)}, {});
+  }
+
+  Expr TransformLayoutInverse(const Expr& expr, const IndexMap& index_map,
+                              const TensorStructInfo& old_tensor_sinfo) {
+    Array<PrimExpr> old_shape = GetShapeFromTensorStructInfo(old_tensor_sinfo);
+    Array<Range> initial_ranges = ConstructRangeFromShape(old_shape);
+    auto [inverse_index_map, padding_predicate] = index_map.NonSurjectiveInverse(initial_ranges);
+    ICHECK(tir::is_zero(padding_predicate))
+        << "Only bijective transformations on input/output buffers are supported, but found "
+           "padding predicate "
+        << padding_predicate << " on initial range " << initial_ranges;
+    return TransformLayout(expr, inverse_index_map);
+  }
+
+  /*!
+   * \brief Adds the \p replacement_func to the module if it has not already been added before.
+   * \returns The global var associated with the PrimFunc.
+   */
+  GlobalVar GetOrCreateGlobalVarForFunc(const PrimFunc& replacement_func, const String& op_kind) {
+    if (cache_.count(replacement_func) != 0) {
+      return cache_[replacement_func];
+    }
+    // Retain the operator name attribute on the replacement PrimFunc. This can help any future
+    // passes that use kOperatorName attribute to identify operator represented by a PrimFunc.
+    PrimFunc replacement_func_with_frozen_layout =
+        WithAttr(replacement_func, kOperatorName, op_kind);
+
+    GlobalVar gv_replacement =
+        builder_->AddFunction(replacement_func_with_frozen_layout, op_kind + "_replacement");
+    cache_.Set(replacement_func, gv_replacement);
+    return gv_replacement;
+  }
+
+  /*!
+   * \brief Updates call_tir args with global var of replacement func and layout transformed inputs
+   */
+  Array<Expr> UpdatedArgs(const GlobalVar& replacement_gv, const Tuple& inputs,
+                          const Array<IndexMap>& transforms) {
+    if (transforms.empty()) return {replacement_gv, inputs};
+
+    Array<Expr> updated_inputs;
+    int index = 0;
+    for (const auto& input : inputs->fields) {
+      auto transform = transforms[index++];
+      ICHECK(IsTransformBijective(input, transform))
+          << "Non bijective transforms on input and output buffers are not supported.";
+      updated_inputs.push_back(TransformLayout(input, transform));
+    }
+    return {replacement_gv, Tuple(updated_inputs)};
```

Comment:
It doesn't seem necessary to pass `replacement_gv` to this function.

Reply:
right, thanks! updated the function